### PR TITLE
feat(storage): json downloads

### PR DESCRIPTION
**Pull request by BrennaEpp**

### CODE REVIEW

File: storage/integration_test.go [hunk 1]
```diff
@@ -253,9 +253,11 @@ func testConfigGRPC(ctx context.Context, t *testing.T, opts ...option.ClientOpti
 
 // initTransportClients initializes Storage clients for each supported transport.
 func initTransportClients(ctx context.Context, t *testing.T, opts ...option.ClientOption) map[string]*Client {
+	withJSON := append(opts, WithJSONReads())
 	return map[string]*Client{
-		"http": testConfig(ctx, t, opts...),
-		"grpc": testConfigGRPC(ctx, t, opts...),
+		"http":      testConfig(ctx, t, opts...),
+		"grpc":      testConfigGRPC(ctx, t, opts...),
+		"jsonReads": testConfig(ctx, t, withJSON...),
```
Comment (tritone) [inline storage/integration_test.go:260]: How much time does this add to the test run? Wondering if it's worth limiting this to reader tests rather than running all transports again.

File: storage/integration_test.go [hunk 2]
```diff
@@ -1334,19 +1337,20 @@ func TestIntegration_Objects(t *testing.T) {
 			if got, want := rc.ContentType(), "text/plain"; got != want {
 				t.Errorf("ContentType (%q) = %q; want %q", obj, got, want)
 			}
-			if got, want := rc.CacheControl(), "public, max-age=60"; got != want {
-				t.Errorf("CacheControl (%q) = %q; want %q", obj, got, want)
-			}
+			// differs
```
Comment (tritone) [inline storage/integration_test.go:1340]: Thoughts on what's going on here? Should we block off time to pair/discuss?

File: storage/storage.go [hunk 3]
```diff
@@ -1704,6 +1710,8 @@ type Conditions struct {
 	// GenerationNotMatch specifies that the object must not have the given
 	// generation for the operation to occur.
 	// If GenerationNotMatch is zero, it has no effect.
+	// This condition only works for object reads if the WithJSONReads client
```
Comment (tritone) [inline storage/storage.go:1713]: `[WithJSONReads]` to make a link, I think

File: storage/storage.go [hunk 4]
```diff
@@ -1722,10 +1730,13 @@ type Conditions struct {
 	// MetagenerationNotMatch specifies that the object must not have the given
 	// metageneration for the operation to occur.
 	// If MetagenerationNotMatch is zero, it has no effect.
+	// This condition only works for object reads if the WithJSONReads client
+	// option is set.
 	MetagenerationNotMatch int64
 }
 
 func (c *Conditions) validate(method string) error {
+
```
Comment (tritone) [inline storage/storage.go:1739]: extra line

File: storage/reader_test.go [hunk 5]
```diff
@@ -387,6 +393,8 @@ func TestContentEncodingGzipWithReader(t *testing.T) {
 			gz := gzip.NewWriter(w)
 			gz.Write(original)
 			gz.Close()
+		default:
+			fmt.Fprintf(w, "unrecognized URL %s", r.URL)
```
Comment (tritone) [inline storage/reader_test.go:397]: This should be a failure?

File: storage/reader.go [hunk 6]
```diff
@@ -142,12 +142,17 @@ func uncompressedByServer(res *http.Response) bool {
 func parseCRC32c(res *http.Response) (uint32, bool) {
 	const prefix = "crc32c="
 	for _, spec := range res.Header["X-Goog-Hash"] {
-		if strings.HasPrefix(spec, prefix) {
-			c, err := decodeUint32(spec[len(prefix):])
-			if err == nil {
-				return c, true
+		values := strings.Split(spec, ",")
```
Comment (tritone) [inline storage/reader.go:145]: Can you speak about why this change is necessary? How does the header format differ between APIs?

File: storage/option.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+// A storageClientOption is an option for a Google Storage client.
+type storageClientOption interface {
+	option.ClientOption
+	ApplyStorageOpt(*storageConfig)
+}
+
+// WithJSONReads is an option that may be passed to a Storage Client on creation.
+// It sets the client to use the JSON API for object reads.
+// Set this option to obtain new features such as the GenerationNotMatch and
```
Comment (tritone) [inline storage/option.go:49]: These aren't really "new" features. You can just say that it is required to use GenerationNotMatch. (MetagenerationMatch shouldn't be relevant for downloads).

File: storage/option.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+// MetagenerationNotMatch conditions.
+//
+// Note that when this option is set, reads will return a zero date for
+// [ReaderObjectAttrs].LastModified and may return a different value for
```
Comment (tritone) [inline storage/option.go:53]: Can you do `[ReaderObjectAttrs.LastModified]` to link to the specific field?

File: storage/option.go [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+// [ReaderObjectAttrs].CacheControl.
+func WithJSONReads() option.ClientOption {
+	return &withReadAPI{useJSON: true}
+}
+
+// WithXMLReads is an option that may be passed to a Storage Client on creation.
+// It sets the client to use the JSON API for object reads.
+//
+// This is the current default, but will be deprecated in the future.
```
Comment (tritone) [inline storage/option.go:62]: I don't think we have to discuss deprecation now.

File: storage/option.go [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+// A storageClientOption is an option for a Google Storage client.
+type storageClientOption interface {
+	option.ClientOption
+	ApplyStorageOpt(*storageConfig)
+}
+
+// WithJSONReads is an option that may be passed to a Storage Client on creation.
+// It sets the client to use the JSON API for object reads.
```
Comment (tritone) [inline storage/option.go:48]: Note here that XML is used by default.

File: storage/http_client.go [hunk 11]
```diff
@@ -777,6 +780,13 @@ func (c *httpStorageClient) NewRangeReader(ctx context.Context, params *newRange
...
 	defer func() { trace.EndSpan(ctx, err) }()
 
+	if c.config.useJSONforReads {
+		return c.newRangeReaderJSON(ctx, params, opts...)
+	}
+	return c.newRangeReaderXML(ctx, params, opts...)
+}
+
+func (c *httpStorageClient) newRangeReaderXML(ctx context.Context, params *newRangeReaderParams, opts ...storageOption) (r *Reader, err error) {
 	s := callSettings(c.settings, opts...)
```
Comment (tritone) [inline storage/http_client.go:790]: Let's make the settings in `NewRangeReader` and just pass it through.

File: storage/http_client.go [hunk 12]
```diff
@@ -975,6 +986,195 @@ func (c *httpStorageClient) NewRangeReader(ctx context.Context, params *newRange
...
+		if params.length < 0 && start < 0 {
+			call.Header().Set("Range", fmt.Sprintf("bytes=%d", start))
+		} else if params.length < 0 && start > 0 {
+			call.Header().Set("Range", fmt.Sprintf("bytes=%d-", start))
+		} else if params.length > 0 {
+			// The end character isn't affected by how many bytes we've seen.
+			call.Header().Set("Range", fmt.Sprintf("bytes=%d-%d", start, params.offset+params.length-1))
+		}
+		// We wait to assign conditions here because the generation number can change in between reopen() runs.
+		if err := applyConds("NewReader", params.gen, params.conds, call); err != nil {
```
Comment (tritone) [inline storage/http_client.go:1026]: We should probably add a test to verify that preconditions including ifGenerationNotMatch are passed through as expected

File: storage/http_client.go [hunk 13]
```diff
@@ -975,6 +986,195 @@ func (c *httpStorageClient) NewRangeReader(ctx context.Context, params *newRange
...
+				var e *googleapi.Error
+				if errors.As(err, &e) {
+					if e.Code == http.StatusNotFound {
+						return ErrObjectNotExist
+					}
+				}
+				return err
+			}
+
+			if res.StatusCode == http.StatusNotFound {
```
Comment (tritone) [inline storage/http_client.go:1047]: Is this not duplicating what's above?

File: storage/http_client.go [hunk 14]
```diff
@@ -975,6 +986,195 @@ func (c *httpStorageClient) NewRangeReader(ctx context.Context, params *newRange
...
+			// back the whole file regardless of the range count passed in as per:
+			//      https://cloud.google.com/storage/docs/transcoding#range,
+			// thus we have to manually move the body forward by seen bytes.
+			if decompressiveTranscoding(res) && seen > 0 {
+				_, _ = io.CopyN(ioutil.Discard, res.Body, seen)
+			}
+
+			// If a generation hasn't been specified, and this is the first response we get, let's record the
+			// generation. In future requests we'll use this generation as a precondition to avoid data races.
+			if params.gen < 0 && res.Header.Get("X-Goog-Generation") != "" {
```
Comment (tritone) [inline storage/http_client.go:1081]: Do we need to get this from the header, or is it populated in the apiary response somewhere?

File: storage/http_client.go [hunk 15]
```diff
@@ -975,6 +986,195 @@ func (c *httpStorageClient) NewRangeReader(ctx context.Context, params *newRange
...
+		Attrs:    attrs,
+		size:     size,
+		remain:   remain,
+		wantCRC:  crc,
+		checkCRC: checkCRC,
+		reader: &httpReader{
+			reopen: reopen,
+			body:   body,
+		},
+	}, nil
```
Comment (tritone) [inline storage/http_client.go:1175]: I'm wondering if it's possible to deduplicate some of this code with what you have for XML. I don't have a great idea of how to do this off the top of my head, but worth thinking about maybe.

File: storage/reader.go (hunk 6 above)
Comment (BrennaEpp) [reply storage/reader.go:148]: I added a comment to the function. 
The XML header looks something like: [crc32c=xxx md5=xxx] 
while the JSON header looks like:         [crc32c=xxx,md5=xxx]

So I added on a split on comma

File: storage/storage.go (hunk 3 above)
Comment (BrennaEpp) [reply storage/storage.go:1718]: I double checked this, it does not create a link in VSCode nor on pkg.go.dev docs. See attachment.

image

File: storage/option.go (hunk 8 above)
Comment (BrennaEpp) [reply storage/option.go:53]: Double-checked this one too, looks like the answer is no:
image

File: storage/option.go (hunk 7 above)
Comment (BrennaEpp) [reply storage/option.go]: Good point, I updated the comment.

File: storage/integration_test.go (hunk 1 above)
Comment (BrennaEpp) [reply storage/integration_test.go:260]: Tests are limited to reader tests using the `skipJSONReads` context.

File: storage/integration_test.go (hunk 2 above)
Comment (BrennaEpp) [reply storage/integration_test.go]: I separated the reader portion out of this test, and tested last modified and cache control separately.

File: storage/reader_test.go (hunk 5 above)
Comment (BrennaEpp) [reply storage/reader_test.go:393]: As discussed offline, the mock just returns this and the test decides how to use it and whether it is a failure state - in this case, the test will fail upon getting such a response.

**Review by tritone — COMMENTED** [review]
Generally looking good, a few more minor things

File: storage/retry_conformance_test.go [hunk 16]
```diff
@@ -247,13 +257,22 @@ var methods = map[string][]retryFunc{
...
-			c.config.useJSONforReads = true
+			client, ok := c.tc.(*httpStorageClient)
+			if ok {
+				client.config.readAPIWasSet = true
+				client.config.useJSONforReads = true
+				defer func() {
+					client.config.readAPIWasSet = false
+					client.config.useJSONforReads = false
+				}()
+			}
```
[the remark this answers is not in the document]
Comment (BrennaEpp) [reply storage/retry_conformance_test.go:275]: hm maybe, did we have a plan for running these tests with GRPC too?

File: storage/reader_test.go [hunk 17]
```diff
@@ -35,42 +35,40 @@ import (
...
-			continue
-		}
-		gotb, err := ioutil.ReadAll(r)
-		if err != nil {
-			t.Errorf("%d/%d: %v", test.offset, test.length, err)
-			continue
-		}
-		if got := string(gotb); got != test.want {
-			t.Errorf("%d/%d: got %q, want %q", test.offset, test.length, got, test.want)
+	multiReaderTest(ctx, t, func(t *testing.T, c *Client) {
```
[the remark this answers is not in the document]
Comment (BrennaEpp) [reply storage/reader_test.go:42]: Just that, and adding the JSON URL to the mock GCS service used in TestContentEncodingGzipWithReader

File: storage/integration_test.go [hunk 18]
```diff
@@ -4053,25 +4098,96 @@ func TestIntegration_ReaderAttrs(t *testing.T) {
...
+			t.Fatalf("cache control; got: %s, want: %s", got, want)
+		}
 	})
 }
 
 // Test that context cancellation correctly stops a download before completion.
 func TestIntegration_ReaderCancel(t *testing.T) {
 	multiTransportTest(context.Background(), t, func(t *testing.T, ctx context.Context, bucket, _ string, client *Client) {
-		ctx, close := context.WithDeadline(ctx, time.Now().Add(time.Second*30))
+		ctx, close := context.WithDeadline(ctx, time.Now().Add(time.Minute))
```
[the remark this answers is not in the document]
Comment (BrennaEpp) [reply storage/integration_test.go]: It was failing locally, I assume due to a very slow connection at the time. Changed it back as it has no longer been an issue.

File: storage/http_client.go (hunk 14 above)
Comment (BrennaEpp) [reply storage/http_client.go]: As we discussed the other day, it indeed must be from the header as  apiary doesn't add it explicitly to the response anywhere.

File: storage/http_client.go (hunk 15 above)
Comment (BrennaEpp) [reply storage/http_client.go]: Gave it a go at splitting it up, PTAL

File: storage/http_client.go (hunk 12 above)
Comment (BrennaEpp) [reply storage/http_client.go]: Added an integration test for the conditions on reader

File: storage/grpc_client.go [hunk 19]
```diff
@@ -861,13 +861,6 @@ func (c *grpcStorageClient) NewRangeReader(ctx context.Context, params *newRange
 		ctx = setUserProjectMetadata(ctx, s.userProject)
 	}
 
-	// A negative length means "read to the end of the object", but the
```
Comment (tritone) [inline storage/grpc_client.go:858]: Is there a reason to make this gRPC changes in the same PR? it might make more sense to do them separately unless necessary.

File: storage/integration_test.go [hunk 20]
```diff
@@ -4203,10 +4206,96 @@ func TestIntegration_ReaderCacheControl(t *testing.T) {
...
+		_, err := o.NewReader(ctx)
+		if !errors.Is(err, ErrObjectNotExist) {
+			t.Fatalf("expected ErrObjectNotExist, got %v", err)
+		}
+	})
+}
+
+// TestIntegration_JSONReaderConditions tests only JSON reads as some conditions
+// do not work with XML.
+func TestIntegration_JSONReaderConditions(t *testing.T) {
```
Comment (tritone) [inline storage/integration_test.go:4222]: Thanks for adding this!

File: storage/http_client.go [hunk 21]
```diff
@@ -804,191 +803,28 @@ func (c *httpStorageClient) newRangeReaderXML(ctx context.Context, params *newRa
...
-		}, s.retry, s.idempotent, setRetryHeaderHTTP(nil))
-		if err != nil {
-			return nil, err
-		}
-		return res, nil
-	}
+	reopen := readerReopen(ctx, req.Header, params, s,
+		func() (*http.Response, error) { return c.hc.Do(req) },
+		func() error { return setConditionsHeaders(req.Header, params.conds) },
+		func(gen int64) { req.URL.RawQuery = fmt.Sprintf("generation=%d", params.gen) })
```
Comment (tritone) [inline storage/http_client.go:818]: Looks like the function `gen int64` arg isn't actually needed/used, you can just remove it for clarity (and use `params.gen` directly for JSON as well).

**Review by tritone — COMMENTED** [review]
Couple more small things, in general the refactor looks good and I'm comfortable going ahead and merging this

File: storage/grpc_client.go (hunk 19 above)
Comment (BrennaEpp) [reply storage/grpc_client.go]: Good point. Originally it was to fix part of the test for GRPC, but we are skipping it anyway. Separate PR opened.

**Review by tritone — APPROVED** [review]
